### ISSUE TITLE
Inference of the same frame multiple times with RTSP stream

### ISSUE THREAD

**Issue by rnoriega** [ultralytics/ultralytics#3559]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection

### Bug

When I iterate over the `model.predict` results with an RTSP stream, YOLO seems to infer over the last image over and over again until it receives a new one (the `result.orig_imag` are the same), generating over 20 FPS inferred with a camera that is limited to 1 FPS. Reading with OpenCV it does read at 1 FPS.

### Environment

- YOLO: Ultralytics YOLOv8.0.128 🚀 Python-3.11.3 torch-2.0.1 CPU
- OS: macOS Ventura 13.4 (22F66)

### Minimal Reproducible Example

```python
from time import monotonic

import cv2
from ultralytics import YOLO

RTSP_URL = 'rtsp://…'
TIMEOUT = 60


def main():
    model = YOLO('yolov8n.pt')
    count = 0
    start = monotonic()

    # 1. YOLO
    # 1/1: rtsp://… Success ✅ (inf frames of shape 1024x576 at 1.00 FPS)
    for result in model.predict(source=RTSP_URL, classes=[0], stream=True, device='mps'):
        count += 1
        if monotonic() - start >= TIMEOUT:
            break

    print(f'{round(count / TIMEOUT, 2)} FPS')  # 24.07 FPS ???

    # 2. OpenCV
    cap = cv2.VideoCapture(RTSP_URL)
    count = 0
    start = monotonic()

    try:
        while monotonic() - start < TIMEOUT:
            cap.read()
            count += 1
    finally:
        cap.release()

    print(f'{round(count / TIMEOUT, 2)} FPS')  # 1.0 FPS


if __name__ == '__main__':
    main()
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@rnoriega interesting. Maybe the RTSP loader is assuming 30FPS and is not using the FPS retrieved by cv2. The loader is here. If you find a fix please submit a PR for this, thank you!

https://github.com/ultralytics/ultralytics/blob/ad99246ff18828273bec2961c6c0a019173fc718/ultralytics/yolo/data/dataloaders/stream_loaders.py#L31-L107

**Comment by glenn-jocher** [comment]
@rnoriega hello there!

After reviewing the code and considering the behavior you described, it seems that the issue you're experiencing is caused by the `__next__` method returning the last fetched images from each stream. However, due to faster inference speeds, the method can be called multiple times before new images are fetched, resulting in the same frame being inferred multiple times until a new one is received.

To address this issue, there are a couple of possible solutions:

1. Waiting for all images to be updated: One solution is to wait until all the images from each stream have been updated before returning them in the `__next__` method. However, this approach may cause the lowest FPS stream to slow down the others if there are multiple streams.

2. Returning only new images: Another solution is to wait until there is at least one new image from any of the streams and return that image along with any other new images. This approach ensures that only the latest frames are used for inference.

3. Iterating cyclically through `self.imgs`: Alternatively, you can iterate cyclically through the `self.imgs` list and return the first updated image, while keeping track of the next index to start from on the next call to `__next__`. This approach ensures that each stream's frames are processed without repeating frames.

Each of these solutions has its trade-offs in terms of the structure of the returned data. The first approach may result in varying lengths of the returned data, while the second and third approaches may not have a fixed length or may have a length different from the number of streams.

I hope this explanation helps! Let me know if you have any further questions or need additional clarification.

**Comment by rnoriega** [comment]
I don't understand answering with my own message but with other words, and I think with ChatGPT.

**Comment by glenn-jocher** [comment]
@rnoriega Thank you for your insights and analysis of the issue. The points you brought up are valid concerns regarding the current implementation of the `__next__` method in the YOLOv8 repository. It seems that the method may return the same frame multiple times until new frames are fetched, resulting in redundant inference.

To address this, one possible solution is to modify the `__next__` method to wait until at least one new image is available before returning it along with any other new images. This ensures that only the latest frames are used for inference. By iterating through the `self.imgs` list cyclically, you can keep track of the next index to start from on the next call to `__next__`, ensuring that each stream's frames are processed without repeating frames.

Alternatively, you could wait for all the frames from each stream to be updated before returning them in the `__next__` method. However, this approach may cause the lowest FPS stream to slow down the others if there are multiple streams.

Both solutions have trade-offs, such as varying lengths of the returned data or data that may not match the number of streams. Ultimately, the decision on which solution to implement depends on the specific use case and requirements.

I hope this explanation clarifies the issue you raised. If you have any further questions or require additional guidance, please feel free to ask.

**Comment by glenn-jocher** [comment]
@rnoriega great news 😃! Your original issue may now be resolved ✅ in PR #3602. This update waits for all streams to present a frame before returning a list of frames, so this will run at the slowest FPS in a group of streams, but will guarantee that no frames are dropped. This can be used together with `vidstride`, i.e.

```
yolo predict show source='https://youtu.be/Zgi9g1ksQHc' imgsz=320

yolo predict show source='https://youtu.be/Zgi9g1ksQHc' imgsz=320 vid_stride=30
``` 

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!

**Comment by rnoriega** [comment]
Sorry for the delay in responding, but yes, now it works as expected. Thank you 😃.

**Comment by glenn-jocher** [comment]
@rnoriega no problem at all! We're glad to hear that the issue has been resolved and that everything is working as expected now. If you have any further questions or run into any other issues, please don't hesitate to ask. We're here to help! Thank you for your patience and support. 😃
